### PR TITLE
Query node transaction failed event

### PR DESCRIPTION
**Pull request by metmirr**
Add an event handler for `TransactionFailed` event, only successful operations are executed by the event handler.

### CODE REVIEW

File: query-node/mappings/content-directory/transaction.ts [hunk 1]
```diff
@@ -81,33 +81,45 @@ async function getNextEntityId(db: DB): Promise<number> {
...
+  debug(`TransactionFailed event: ${JSON.stringify(event)}`)
+
+  const failedOperationIndex = event.params[1].value as number
+  const operations = decode.getOperations(event)
+
+  if (operations.length === 0 || operations.length === 1) return
+
+  const successfulOperations = operations.filter((op, index) => index < failedOperationIndex)
+
+  const {
```
Comment (mnaamani) [inline query-node/mappings/content-directory/transaction.ts]: You can factor out this common code in `contentDirectory_TransactionFailed` and `contentDirectory_TransactionCompleted` into a function `applyOperations(operations: IBatchOperation)`

File: query-node/mappings/content-directory/transaction.ts [hunk 2]
```diff
@@ -81,33 +81,45 @@ async function getNextEntityId(db: DB): Promise<number> {
...
+// eslint-disable-next-line @typescript-eslint/naming-convention
+export async function contentDirectory_TransactionFailed(db: DB, event: SubstrateEvent): Promise<void> {
+  debug(`TransactionFailed event: ${JSON.stringify(event)}`)
+
+  const failedOperationIndex = event.params[1].value as number
+  const operations = decode.getOperations(event)
+
+  if (operations.length === 0 || operations.length === 1) return
+
+  const successfulOperations = operations.filter((op, index) => index < failedOperationIndex)
```
Comment (mnaamani) [inline query-node/mappings/content-directory/transaction.ts]: I think you can combine these two statements into a single one using `operations.slice()`

**Review by mnaamani — CHANGES_REQUESTED** [review]
Implementation looks correct, just left some refactoring change request.
Will test queries now and report back.

File: query-node/mappings/content-directory/transaction.ts (hunk 2 above)
Comment (metmirr) [reply query-node/mappings/content-directory/transaction.ts]: Much better now :) done.

Comment (metmirr) [reply query-node/mappings/content-directory/transaction.ts]: Earlier I attempted to use `slice` but I was getting `TypeError: registry.createClass is not a function`, I couldn't recall calling `toArray` on polkadot codec.

File: query-node/mappings/content-directory/transaction.ts (hunk 1 above)
Comment (metmirr) [reply query-node/mappings/content-directory/transaction.ts]: Done.

File: query-node/mappings/content-directory/transaction.ts [hunk 3]
```diff
@@ -88,19 +89,10 @@ export async function contentDirectory_TransactionFailed(db: DB, event: Substrat
   const failedOperationIndex = event.params[1].value as number
   const operations = decode.getOperations(event)
 
-  if (operations.length === 0 || operations.length === 1) return
+  const successfulOperations = operations.toArray().slice(2, failedOperationIndex)
```
Comment (mnaamani) [inline query-node/mappings/content-directory/transaction.ts]: why are we always skipping first two operations?

Comment (metmirr) [reply query-node/mappings/content-directory/transaction.ts]: If `operations` length is `0` then there is no successful operation, if the length is `1` then there is only one operation which is the failed one (that's why the `TransactionFailed` event is triggered for). In both cases, there is nothing for me to process.

Comment (metmirr) [reply query-node/mappings/content-directory/transaction.ts]: Okay, that's wrong! I am updating now! I mixed the indices with the length

**Pull request merged**